catlas/parity/data_processing_utils.py

```diff
@@ -0,0 +1,127 @@
+        return energy[0]
+
+    @staticmethod
+    def _get_bulk_elements_and_num(stoichiometry):
+        """Get the unique bulk elements and number of from stoichiometry"""
+        return list(stoichiometry.keys()), len(list(stoichiometry.keys()))
+
+    def process_data(self):
+        """Creates the primary dataframe for use in analysis"""
+
+        model_id = self.npz_path.split("/")[-1]
+        model_id = model_id.split(".")[0]
+
+        df_path = "catlas/parity/df_pkls/" + model_id + ".pkl"
+
+        # Open files
+        ML_data = dict(load(self.npz_path))
+        dft_df = pd.read_pickle(self.dft_df_path)
+
+        # Get ML energies
+        dft_df["ML_energy"] = dft_df.apply(
+            self._get_predicted_E, args=(ML_data,), axis=1
+        )
+        dft_df["bulk_elements"], dft_df["bulk_nelements"] = zip(
+            *dft_df.stoichiometry.apply(self._get_bulk_elements_and_num)
+        )
+
+        # Pickle results for future use
+        dft_df.to_pickle(df_path)
+
+
+class ProcessValTraj:
+    def __init__(self, traj_path):
+        """
+        Initialize class to handle trajectory analysis
+
+        Args:
+            traj_path: the path to the trajectory to be considered
+        """
+        self.traj_path = traj_path
+
+    def _get_status(self, traj, frame):
+        """Inspects the trajectory for anomolies."""
+        try:
+            status = {}
+            detector = DetectTrajAnomaly(traj[0], traj[frame], traj[0].get_tags())
+            status["dissociation"] = detector.is_adsorbate_dissociated()
+            status["desorption"] = detector.is_adsorbate_desorbed(neighbor_thres=3)
+            status["reconstruction"] = detector.is_surface_reconstructed(
+                slab_movement_thres=1
+            )
+            return all(
+                [
+                    ~status["dissociation"],
+                    ~status["desorption"],
+                    ~status["reconstruction"],
+                ]
+            )
+
+        except:
+            return False
+
+    def _get_energies(self, traj):
+        """Iterate through the trajectory frame and return per frame energies."""
+        try:
+            energies = []
+            for frame in traj:
+                energy = frame.get_potential_energy()
+                energies.append(energy)
+            return energies
+        except:
+            return [np.nan] * len(traj)
+
+    def get_result(self):
+        name = self.traj_path.split("/")[-1].split(".")[0]
+        random_num = name.split("/")[-1]
+        random_id = "random" + str(random_num)
+        try:
+            traj = Trajectory(self.traj_path)
+            status = self._get_status(traj, -1)
+            energy = self._get_energies(traj)
+            now = {"random_id": random_id, "ML_energy": energy, "status": status}
+            return now
+
+        except:
+            return {"random_id": random_id, "ML_energy": "failed", "status": "failed"}
```

Comment:
Maybe add the exception as the key for status?

Reply:
changed this to be boolean because I think that makes more sense. I think this will basically never happen. Its just in case!